### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by alexeygrigorev** [alexeygrigorev/libffm-python#1]
by @4yu:

sorry,i can't find the issue link,just write here. when i try this in my online service: 

```
ffm_data = ffm.FFMData(arr_feature, y) 
predictions = model.predict(ffm_data);
```

then,the server's memory increase continuously.the problem perhaps come from ffm.cpp `ffm_convert_data` function.

please have a look,thank you.

**Comment by alexeygrigorev** [comment]
@4yu thanks for submitting the issue. 
Pull requests with the solution are welcome

**Comment by 4yu** [comment]
sorry,i am not familiar with c++.
i just trace the python program step by step ,and found the problem is in ffm.cpp ffm_convert_data function.[when i don't invoke the ffm_convert_data  function,the memory is stable]

**Comment by alexklibisz** [comment]
Indeed it seems to be in `ffm_convert_data`. Here is a simple script that reproduces the memory leak.

```
# git clone https://github.com/alexeygrigorev/libffm-python.git
# cd libffm-python
# make so
# mv libffm.so ffm
# python setup.py install

import ffm
import numpy as np
import pdb
import resource


def mem(prefix):
    print('%s %s (kb)' % (prefix, resource.getrusage(resource.RUSAGE_SELF).ru_maxrss))

X = [[(1, 2, 1), (2, 3, 1), (3, 5, 1)]] * 10000
y = np.random.rand(10000).round()

model = ffm.FFM(eta=0.1, lam=0.0001, k=4)

# Say you want to train the same model several times
# with different data (e.g. cross-validation)

for i in range(3):

    mem('Before ffm.FFMData(X,y)')
    ffm_data = ffm.FFMData(X, y)
    mem('After ffm.FFMData(X,y)')
    model.init_model(ffm_data)
    mem('After model.init_model(ffm_data)')
    model.iteration(ffm_data)
    mem('After model.iteration(ffm_data)')
    print('*' * 10)
```

I'm currently working to try to fix this.

**Comment by alexeygrigorev** [comment]
I suspect another reason may be that the array for predictions is created in the library, and thus not managed by python's gc. 

I'm not sure, but maybe creating the output array outside of the library, e.g. with numpy, could solve the problem.

**Comment by alexeygrigorev** [comment]
@alexklibisz in your code there's a problem: init model should be executed only once, before the loop, otherwise you're basically resetting the model's weights at each iteration. But it doesn't mean it doesn't cause the leak - probably the model struct, including all the weights, should also be managed by python

**Comment by alexeygrigorev** [comment]
Probably it's also true for `init_model` - maybe it's not managed by python now, so gc can't collect it when no longer needed.

**Comment by alexeygrigorev** [comment]
Probably it's also true for `convert_data` - maybe it's not managed by python now, so gc can't collect it when no longer needed.

**Comment by alexklibisz** [comment]
I resolved the problem and will post the fix later today. When you make a
dataset new memory is allocated but never freed.

**Comment by alexeygrigorev** [comment]
addressed in #3
